# SQL Lab

In this lab we will learn how to use execute SQL from the iPython notebook and practice some queries on the [Northwind sample database](https://northwinddatabase.codeplex.com/).

You can access the data with this command:

    psql -h dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com -p 5432 -U dsi_student northwind
    password: gastudents


First of all let's install the ipython-sql extension. You can find instructions [here](https://github.com/catherinedevlin/ipython-sql).

In [1]:
# Necessary package 1 (if not already installed)
!pip install ipython-sql

In [2]:
# Necessary package 2 (if not already installed)
!pip install psycopg2

Let's see if it works:

In [3]:
# Magic Function Load Extension
%load_ext sql

In [4]:
# Line Magic Function SQL Extension and path w/ access info to datafile

In [5]:
%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind

'Connected: dsi_student@northwind'

In [6]:
# This code will not run. Unfortunately, it seems that you can't run cell magic in cells that have comments. This
# does make sense, given that # are not the SQL syntax for commenting. Also, SQL comments don't seem to work
# in Jupyter notebook. Weird.

%%sql
SELECT * 
FROM orders 
LIMIT 5;

SyntaxError: invalid syntax (<ipython-input-6-9a00858c8bbe>, line 5)

In [ ]:
# So try to comment in cells above or below your code. Either use comments, like this cell...

...or create a Markdown cell, like this cell!

In [7]:
%%sql
SELECT * 
FROM orders 
LIMIT 5;

5 rows affected.


[(10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France'),
 (10249, 'TOMSP', 6, datetime.date(1996, 7, 5), datetime.date(1996, 8, 16), datetime.date(1996, 7, 10), 1, 11.61, 'Toms Spezialitäten', 'Luisenstr. 48', 'Münster', None, '44087', 'Germany'),
 (10250, 'HANAR', 4, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 12), 2, 65.83, 'Hanari Carnes', 'Rua do Paço, 67', 'Rio de Janeiro', 'RJ', '05454-876', 'Brazil'),
 (10251, 'VICTE', 3, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 15), 1, 41.34, 'Victuailles en stock', '2, rue du Commerce', 'Lyon', None, '69004', 'France'),
 (10252, 'SUPRD', 4, datetime.date(1996, 7, 9), datetime.date(1996, 8, 6), datetime.date(1996, 7, 11), 2, 51.3, 'Suprêmes délices', 'Boulevard Tirou, 255', 'Charleroi', None, 'B-6000', 'Belgium')]

Nice!!! We can now go ahead with the lab!

## 1: Inspect the database

If we were connected via console, it would be easy to list all tables using `\dt`. We can however access table information performing a query on the `information_schema.tables` table.

### 1.a: List Tables

1. write a `SELECT` statement that lists all the tables in the public schema of the `northwind` database, sorted alphabetically

In [ ]:
# %%SQL earlier picked our postgres filepath so we just have to call %%sql to reach our DB

In [8]:
%%sql
SELECT * FROM information_schema.tables

217 rows affected.


[('northwind', 'pg_catalog', 'pg_type', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('northwind', 'public', 'categories', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_roles', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_group', 'VIEW', None, None, None, None, None, 'YES', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_user', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_rules', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_views', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_tables', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_matviews', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_indexes', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_stats', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_settings', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_locks', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_cursors', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_available_extensions', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_available_extension_versions', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_prepared_xacts', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_prepared_statements', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_seclabels', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_timezone_abbrevs', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_timezone_names', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_stat_all_tables', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_stat_xact_all_tables', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_stat_sys_tables', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_stat_xact_sys_tables', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_stat_user_tables', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_stat_xact_user_tables', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_statio_all_tables', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_statio_sys_tables', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_statio_user_tables', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_attribute', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_proc', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_statio_all_indexes', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_statio_sys_indexes', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_statio_user_indexes', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_statio_all_sequences', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_statio_sys_sequences', 'VIEW', None, None, None, None, None, 'NO', 'NO', None),
 ('northwind', 'pg_catalog', 'pg_statio_user_sequences', 'VIEW', N

### 1.b: Print Schemas

The table `INFORMATION_SCHEMA.COLUMNS` contains schema information on each.

Query it to display schemas of all the public tables. In particular we are interested in the column names and data types. Make sure you only include public schemas to avoid cluttering your results with a bunch of postgres related stuff.

In [ ]:
# Information_schema is like an automatically created and updated table of tables

In [9]:
%%sql
SELECT * FROM information_schema.columns

1825 rows affected.


[('northwind', 'information_schema', 'role_table_grants', 'grantee', 2, None, 'YES', 'character varying', None, 1073741824, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'information_schema', 'sql_identifier', 'northwind', 'pg_catalog', 'varchar', None, None, None, None, '2', 'NO', 'NO', None, None, None, None, None, None, 'NEVER', None, 'NO'),
 ('northwind', 'information_schema', 'parameters', 'udt_schema', 25, None, 'YES', 'character varying', None, 1073741824, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'information_schema', 'sql_identifier', 'northwind', 'pg_catalog', 'varchar', None, None, None, None, '25', 'NO', 'NO', None, None, None, None, None, None, 'NEVER', None, 'NO'),
 ('northwind', 'pg_catalog', 'pg_user_mappings', 'srvname', 3, None, 'YES', 'name', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'pg_catalog', 'name', None, None, None, None, '3', 'NO', 'NO', None, None, None, None, None, None, 'NEVER', None, 'NO'),
 ('northwind', 'pg_catalog', 'pg_type', 'typarray', 13, None, 'NO', 'oid', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'pg_catalog', 'oid', None, None, None, None, '13', 'NO', 'NO', None, None, None, None, None, None, 'NEVER', None, 'YES'),
 ('northwind', 'pg_catalog', 'pg_language', 'lanplcallfoid', 5, None, 'NO', 'oid', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'pg_catalog', 'oid', None, None, None, None, '5', 'NO', 'NO', None, None, None, None, None, None, 'NEVER', None, 'YES'),
 ('northwind', 'pg_catalog', 'pg_class', 'relkind', 16, None, 'NO', '"char"', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'pg_catalog', 'char', None, None, None, None, '16', 'NO', 'NO', None, None, None, None, None, None, 'NEVER', None, 'YES'),
 ('northwind', 'information_schema', 'tables', 'table_type', 4, None, 'YES', 'character varying', None, 1073741824, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'information_schema', 'character_data', 'northwind', 'pg_catalog', 'varchar', None, None, None, None, '4', 'NO', 'NO', None, None, None, None, None, None, 'NEVER', None, 'NO'),
 ('northwind', 'pg_catalog', 'pg_pltemplate', 'tmplname', 1, None, 'NO', 'name', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'pg_catalog', 'name', None, None, None, None, '1', 'NO', 'NO', None, None, None, None, None, None, 'NEVER', None, 'YES'),
 ('northwind', 'pg_catalog', 'pg_cursors', 'is_holdable', 3, None, 'YES', 'boolean', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'pg_catalog', 'bool', None, None, None, None, '3', 'NO', 'NO', None, None, None, None, None, None, 'NEVER', None, 'NO'),
 ('northwind', 'public', 'shippers', 'CompanyName', 2, None, 'NO', 'character varying', 40, 160, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'pg_catalog', 'varchar', None, None, None, None, '2', 'NO', 'NO', None, None, None, None, None, None, 'NEVER', None, 'YES'),
 ('northwind', 'pg_catalog', 'pg_index', 'indisreplident', 13, None, 'NO', 'boolean', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'pg_catalog', 'bool', None, None, None, None, '13', 'NO', 'NO', None, None, None, None, None, None, 'NEVER', None, 'YES'),
 ('northwind', 'information_schema', 'routine_privileges', 'grantee', 2, None, 'YES', 'character varying', None, 1073741824, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'information_schema', 'sql_identifier', 'northwind', 'pg_catalog', 'varchar', None, Non

### 1.c: Table peek

Another way of quickly looking at table information is to query the first few rows. Do this for a couple of tables, for example: `orders`, `products`, `usstates`. Display only the first 3 rows.


In [10]:
categories = %sql SELECT * FROM categories LIMIT 3;
categories = categories.DataFrame()

3 rows affected.


In [11]:
categories

CategoryID CategoryName                                        Description  \
0           1    Beverages        Soft drinks, coffees, teas, beers, and ales   
1           2   Condiments  Sweet and savory sauces, relishes, spreads, an...   
2           3  Confections                Desserts, candies, and sweet breads   

  Picture  
0      []  
1      []  
2      []

In [12]:
orders = %sql SELECT * FROM orders LIMIT 3;
orders = orders.DataFrame()
orders

3 rows affected.


OrderID CustomerID  EmployeeID   OrderDate RequiredDate ShippedDate  \
0    10248      VINET           5  1996-07-04   1996-08-01  1996-07-16   
1    10249      TOMSP           6  1996-07-05   1996-08-16  1996-07-10   
2    10250      HANAR           4  1996-07-08   1996-08-05  1996-07-12   

   ShipVia  Freight                   ShipName         ShipAddress  \
0        3    32.38  Vins et alcools Chevalier  59 rue de l'Abbaye   
1        1    11.61         Toms Spezialitäten       Luisenstr. 48   
2        2    65.83              Hanari Carnes     Rua do Paço, 67   

         ShipCity ShipRegion ShipPostalCode ShipCountry  
0           Reims       None          51100      France  
1         Münster       None          44087     Germany  
2  Rio de Janeiro         RJ      05454-876      Brazil

In [13]:
products = %sql SELECT * FROM products LIMIT 3;
products = products.DataFrame()
products

3 rows affected.


ProductID    ProductName  SupplierID  CategoryID      QuantityPerUnit  \
0          1           Chai           8           1   10 boxes x 30 bags   
1          2          Chang           1           1   24 - 12 oz bottles   
2          3  Aniseed Syrup           1           2  12 - 550 ml bottles   

   UnitPrice  UnitsInStock  UnitsOnOrder  ReorderLevel  Discontinued  
0       18.0            39             0            10             1  
1       19.0            17            40            25             1  
2       10.0            13            70            25             0

As you can see, some tables (like `usstates` or `region`) contain information that is probably less prone to change than other tables (like `orders` or `order_details`). This database is well organized to avoid unnecessary duplication. Let's start digging deeper in the data.

## 2: Products

What products is this company selling? The `products` and `categories` tables contain information to answer this question.

### 2.a: What categories of products is the company selling?

Remember that PostgreSQL is case sensitive.

In [14]:
%%sql
SELECT COUNT(*), "CategoryName" FROM categories 
    GROUP BY "CategoryName"

8 rows affected.


[(1, 'Beverages'),
 (1, 'Produce'),
 (1, 'Seafood'),
 (1, 'Grains/Cereals'),
 (1, 'Confections'),
 (1, 'Meat/Poultry'),
 (1, 'Condiments'),
 (1, 'Dairy Products')]

### 2.b: How many products per category ID does the catalog contain?

In [15]:
%%sql
SELECT COUNT(DISTINCT "ProductID"), "CategoryID" FROM products 
    GROUP BY "CategoryID"
    ORDER BY "CategoryID"

8 rows affected.


[(12, 1), (12, 2), (13, 3), (10, 4), (7, 5), (6, 6), (5, 7), (12, 8)]

### 2.c: How many NOT discontinued products per categoryID? 

In [16]:
%%sql
SELECT COUNT(DISTINCT "ProductID"), "CategoryID" FROM products
    WHERE "Discontinued" = 0
    GROUP BY "CategoryID"
    ORDER BY "CategoryID"

8 rows affected.


[(9, 1), (11, 2), (13, 3), (10, 4), (6, 5), (2, 6), (4, 7), (12, 8)]

### 2.d: What are the 5 most expensive products that were not discontinued?

In [17]:
%%sql
SELECT "ProductID", "ProductName", "UnitPrice" FROM products
    ORDER BY "UnitPrice" DESC
    LIMIT 5

5 rows affected.


[(38, 'Côte de Blaye', 263.5),
 (29, 'Thüringer Rostbratwurst', 123.79),
 (9, 'Mishi Kobe Niku', 97.0),
 (20, "Sir Rodney's Marmalade", 81.0),
 (18, 'Carnarvon Tigers', 62.5)]

### 2.e: How many units of each of these 5 products are there in stock?

In [18]:
%%sql
SELECT "ProductID", "ProductName", "UnitPrice", "UnitsInStock" FROM products
    ORDER BY "UnitPrice" DESC
    LIMIT 5

5 rows affected.


[(38, 'Côte de Blaye', 263.5, 17),
 (29, 'Thüringer Rostbratwurst', 123.79, 0),
 (9, 'Mishi Kobe Niku', 97.0, 29),
 (20, "Sir Rodney's Marmalade", 81.0, 40),
 (18, 'Carnarvon Tigers', 62.5, 42)]

## 3: Orders

Now that we have a better understanding of products, let's start digging into orders.

In [19]:
orders = %sql SELECT * FROM orders LIMIT 3;
orders = orders.DataFrame()
orders

3 rows affected.


OrderID CustomerID  EmployeeID   OrderDate RequiredDate ShippedDate  \
0    10248      VINET           5  1996-07-04   1996-08-01  1996-07-16   
1    10249      TOMSP           6  1996-07-05   1996-08-16  1996-07-10   
2    10250      HANAR           4  1996-07-08   1996-08-05  1996-07-12   

   ShipVia  Freight                   ShipName         ShipAddress  \
0        3    32.38  Vins et alcools Chevalier  59 rue de l'Abbaye   
1        1    11.61         Toms Spezialitäten       Luisenstr. 48   
2        2    65.83              Hanari Carnes     Rua do Paço, 67   

         ShipCity ShipRegion ShipPostalCode ShipCountry  
0           Reims       None          51100      France  
1         Münster       None          44087     Germany  
2  Rio de Janeiro         RJ      05454-876      Brazil

### 3.a: How many orders in total?

In [20]:
%%sql
SELECT COUNT(*) FROM orders

1 rows affected.


[(830,)]

### 3.b: Which country is receiving the most orders?

In [21]:
%%sql
SELECT COUNT(*), "ShipCountry" FROM orders
    GROUP BY "ShipCountry"
    ORDER BY COUNT(*) DESC
    LIMIT 5

5 rows affected.


[(122, 'USA'), (122, 'Germany'), (83, 'Brazil'), (77, 'France'), (56, 'UK')]

Answer: USA and Germany receive the most orders, with each receiving 122 orders.

### 3.c: Which country is receiving the least?

In [22]:
%%sql
SELECT COUNT(*), "ShipCountry" FROM orders
    GROUP BY "ShipCountry"
    ORDER BY COUNT(*) ASC
    LIMIT 5

5 rows affected.


[(6, 'Norway'),
 (7, 'Poland'),
 (13, 'Portugal'),
 (16, 'Argentina'),
 (18, 'Denmark')]

Answer: Norway receives the fewest orders with 6 orders.

### 3.d: What's the average shipping time (ShippedDate - OrderDate)?

Hint: use an aggregate function

In [23]:
%%sql
SELECT AVG("ShippedDate" - "OrderDate") FROM orders

1 rows affected.


[(Decimal('8.4919653893695921'),)]

### 3.e: What customer is submitting the highest number of orders?

In [24]:
%%sql
SELECT COUNT(*), "CustomerID" FROM orders
    GROUP BY "CustomerID"
    ORDER BY COUNT(*) DESC
    LIMIT 5

5 rows affected.


[(31, 'SAVEA'), (30, 'ERNSH'), (28, 'QUICK'), (19, 'HUNGO'), (19, 'FOLKO')]

Answer: Customer SAVEA is submitting the highest number of orders.

## Shifting from SQL to Pandas

It's likely that, throughout the course, we'll be more familiar with Pandas syntax than SQL. Rather than racking our brains to execute complicated SQL queries when Pandas can answer this for us, what if we query data from SQL and read it into Pandas, then answer our questions that way?

In [25]:
import pandas as pd

The below cell was run earlier, where we saved the output of the SQL query as a variable `categories`. Examine the output.

In [26]:
categories = %sql SELECT "CategoryID", "CategoryName", "Description" FROM categories;
categories

8 rows affected.


[(1, 'Beverages', 'Soft drinks, coffees, teas, beers, and ales'),
 (2, 'Condiments', 'Sweet and savory sauces, relishes, spreads, and seasonings'),
 (3, 'Confections', 'Desserts, candies, and sweet breads'),
 (4, 'Dairy Products', 'Cheeses'),
 (5, 'Grains/Cereals', 'Breads, crackers, pasta, and cereal'),
 (6, 'Meat/Poultry', 'Prepared meats'),
 (7, 'Produce', 'Dried fruit and bean curd'),
 (8, 'Seafood', 'Seaweed and fish')]

In [27]:
categories['CategoryName']

KeyError: 'CategoryName'

We can't directly reference that column in the same way we'd usually like. Let's read this into Pandas! Note that we pass the results of a query into a variable `categories`, then force `categories` to be a Pandas DataFrame by using the `.DataFrame()` method.

In [28]:
categories = %sql SELECT * FROM categories;
categories = categories.DataFrame()

8 rows affected.


In [29]:
categories

CategoryID    CategoryName  \
0           1       Beverages   
1           2      Condiments   
2           3     Confections   
3           4  Dairy Products   
4           5  Grains/Cereals   
5           6    Meat/Poultry   
6           7         Produce   
7           8         Seafood   

                                         Description Picture  
0        Soft drinks, coffees, teas, beers, and ales      []  
1  Sweet and savory sauces, relishes, spreads, an...      []  
2                Desserts, candies, and sweet breads      []  
3                                            Cheeses      []  
4                Breads, crackers, pasta, and cereal      []  
5                                     Prepared meats      []  
6                          Dried fruit and bean curd      []  
7                                   Seaweed and fish      []

#### Let's recreate our answers from above by using Pandas! Make sure to show us the commands you're running to get the answers.

You may find [this link](http://pandas.pydata.org/pandas-docs/stable/comparison_with_sql.html#compare-with-sql-join) particularly helpful!

## 4: Products

What products is this company selling? The `products` and `categories` tables contain information to answer this question.

In [30]:
categories = %sql SELECT * FROM categories;
categories = categories.DataFrame()
categories

8 rows affected.


CategoryID    CategoryName  \
0           1       Beverages   
1           2      Condiments   
2           3     Confections   
3           4  Dairy Products   
4           5  Grains/Cereals   
5           6    Meat/Poultry   
6           7         Produce   
7           8         Seafood   

                                         Description Picture  
0        Soft drinks, coffees, teas, beers, and ales      []  
1  Sweet and savory sauces, relishes, spreads, an...      []  
2                Desserts, candies, and sweet breads      []  
3                                            Cheeses      []  
4                Breads, crackers, pasta, and cereal      []  
5                                     Prepared meats      []  
6                          Dried fruit and bean curd      []  
7                                   Seaweed and fish      []

In [31]:
products = %sql SELECT * FROM products;
products = products.DataFrame()
products

77 rows affected.


ProductID                       ProductName  SupplierID  CategoryID  \
0           1                              Chai           8           1   
1           2                             Chang           1           1   
2           3                     Aniseed Syrup           1           2   
3           4      Chef Anton's Cajun Seasoning           2           2   
4           5            Chef Anton's Gumbo Mix           2           2   
5           6      Grandma's Boysenberry Spread           3           2   
6           7   Uncle Bob's Organic Dried Pears           3           7   
7           8        Northwoods Cranberry Sauce           3           2   
8           9                   Mishi Kobe Niku           4           6   
9          10                             Ikura           4           8   
10         11                    Queso Cabrales           5           4   
11         12         Queso Manchego La Pastora           5           4   
12         13                             Konbu           6           8   
13         14                              Tofu           6           7   
14         15                      Genen Shouyu           6           2   
15         16                           Pavlova           7           3   
16         17                      Alice Mutton           7           6   
17         18                  Carnarvon Tigers           7           8   
18         19        Teatime Chocolate Biscuits           8           3   
19         20            Sir Rodney's Marmalade           8           3   
20         21               Sir Rodney's Scones           8           3   
21         22               Gustaf's Knäckebröd           9           5   
22         23                          Tunnbröd           9           5   
23         24                Guaraná Fantástica          10           1   
24         25           NuNuCa Nuß-Nougat-Creme          11           3   
25         26               Gumbär Gummibärchen          11           3   
26         27                Schoggi Schokolade          11           3   
27         28                 Rössle Sauerkraut          12           7   
28         29           Thüringer Rostbratwurst          12           6   
29         30             Nord-Ost Matjeshering          13           8   
..        ...                               ...         ...         ...   
47         48                         Chocolade          22           3   
48         49                          Maxilaku          23           3   
49         50                  Valkoinen suklaa          23           3   
50         51             Manjimup Dried Apples          24           7   
51         52                          Filo Mix          24           5   
52         53                     Perth Pasties          24           6   
53         54                         Tourtière          25           6   
54         55                      Pâté chinois          25           6   
55         56            Gnocchi di nonna Alice          26           5   
56         57                    Ravioli Angelo          26           5   
57         58            Escargots de Bourgogne          27           8   
58         59              Raclette Courdavault          28           4   
59         60                 Camembert Pierrot          28           4   
60         61                    Sirop d'érable          29           2   
61         62                    Tarte au sucre          29           3   
62         63                      Vegie-spread           7           2   
63         64         Wimmers gute Semmelknödel          12           5   
64         65  Louisiana Fiery Hot Pepper Sauce           2           2   
65         66         Louisiana Hot Spiced Okra           2           2   
66         67         Laughing Lumberjack Lager          16           1   
67         68               Scottish Longbreads           8           3   
68         69                

### 4.a: What categories of products is the company selling?

In [32]:
categories.groupby(by='CategoryName').sum()

CategoryID
CategoryName              
Beverages                1
Condiments               2
Confections              3
Dairy Products           4
Grains/Cereals           5
Meat/Poultry             6
Produce                  7
Seafood                  8

### 4.b: How many products per category ID does the catalog contain?

In [33]:
products

ProductID                       ProductName  SupplierID  CategoryID  \
0           1                              Chai           8           1   
1           2                             Chang           1           1   
2           3                     Aniseed Syrup           1           2   
3           4      Chef Anton's Cajun Seasoning           2           2   
4           5            Chef Anton's Gumbo Mix           2           2   
5           6      Grandma's Boysenberry Spread           3           2   
6           7   Uncle Bob's Organic Dried Pears           3           7   
7           8        Northwoods Cranberry Sauce           3           2   
8           9                   Mishi Kobe Niku           4           6   
9          10                             Ikura           4           8   
10         11                    Queso Cabrales           5           4   
11         12         Queso Manchego La Pastora           5           4   
12         13                             Konbu           6           8   
13         14                              Tofu           6           7   
14         15                      Genen Shouyu           6           2   
15         16                           Pavlova           7           3   
16         17                      Alice Mutton           7           6   
17         18                  Carnarvon Tigers           7           8   
18         19        Teatime Chocolate Biscuits           8           3   
19         20            Sir Rodney's Marmalade           8           3   
20         21               Sir Rodney's Scones           8           3   
21         22               Gustaf's Knäckebröd           9           5   
22         23                          Tunnbröd           9           5   
23         24                Guaraná Fantástica          10           1   
24         25           NuNuCa Nuß-Nougat-Creme          11           3   
25         26               Gumbär Gummibärchen          11           3   
26         27                Schoggi Schokolade          11           3   
27         28                 Rössle Sauerkraut          12           7   
28         29           Thüringer Rostbratwurst          12           6   
29         30             Nord-Ost Matjeshering          13           8   
..        ...                               ...         ...         ...   
47         48                         Chocolade          22           3   
48         49                          Maxilaku          23           3   
49         50                  Valkoinen suklaa          23           3   
50         51             Manjimup Dried Apples          24           7   
51         52                          Filo Mix          24           5   
52         53                     Perth Pasties          24           6   
53         54                         Tourtière          25           6   
54         55                      Pâté chinois          25           6   
55         56            Gnocchi di nonna Alice          26           5   
56         57                    Ravioli Angelo          26           5   
57         58            Escargots de Bourgogne          27           8   
58         59              Raclette Courdavault          28           4   
59         60                 Camembert Pierrot          28           4   
60         61                    Sirop d'érable          29           2   
61         62                    Tarte au sucre          29           3   
62         63                      Vegie-spread           7           2   
63         64         Wimmers gute Semmelknödel          12           5   
64         65  Louisiana Fiery Hot Pepper Sauce           2           2   
65         66         Louisiana Hot Spiced Okra           2           2   
66         67         Laughing Lumberjack Lager          16           1   
67         68               Scottish Longbreads           8           3   
68         69                

In [34]:
products[['CategoryID','ProductID']].groupby(by = 'CategoryID').count()

ProductID
CategoryID           
1                  12
2                  12
3                  13
4                  10
5                   7
6                   6
7                   5
8                  12

### 4.c: How many NOT discontinued products per categoryID? 

In [35]:
products[products['Discontinued'] == 0][['CategoryID','ProductID']].groupby(by = 'CategoryID').count()

ProductID
CategoryID           
1                   9
2                  11
3                  13
4                  10
5                   6
6                   2
7                   4
8                  12

### 4.d: What are the 5 most expensive products that were not discontinued?

In [36]:
products[['ProductName', 'ProductID', 'UnitPrice']][products['Discontinued'] == 0].sort_values(by = 'UnitPrice', ascending = False).head()

ProductName  ProductID  UnitPrice
37           Côte de Blaye         38      263.5
19  Sir Rodney's Marmalade         20       81.0
17        Carnarvon Tigers         18       62.5
58    Raclette Courdavault         59       55.0
50   Manjimup Dried Apples         51       53.0

### 4.e: How many units of each of these 5 products are there in stock?

In [37]:
products[['ProductName', 'ProductID', 'UnitPrice', 'UnitsInStock']][products['Discontinued'] == 0].sort_values(by = 'UnitPrice', ascending = False).head()

ProductName  ProductID  UnitPrice  UnitsInStock
37           Côte de Blaye         38      263.5            17
19  Sir Rodney's Marmalade         20       81.0            40
17        Carnarvon Tigers         18       62.5            42
58    Raclette Courdavault         59       55.0            79
50   Manjimup Dried Apples         51       53.0            20

## 5: Orders

Now that we have a better understanding of products, let's start digging into orders.

In [38]:
orders = %sql SELECT * FROM orders;
orders = orders.DataFrame()
orders

830 rows affected.


OrderID CustomerID  EmployeeID   OrderDate RequiredDate ShippedDate  \
0      10248      VINET           5  1996-07-04   1996-08-01  1996-07-16   
1      10249      TOMSP           6  1996-07-05   1996-08-16  1996-07-10   
2      10250      HANAR           4  1996-07-08   1996-08-05  1996-07-12   
3      10251      VICTE           3  1996-07-08   1996-08-05  1996-07-15   
4      10252      SUPRD           4  1996-07-09   1996-08-06  1996-07-11   
5      10253      HANAR           3  1996-07-10   1996-07-24  1996-07-16   
6      10254      CHOPS           5  1996-07-11   1996-08-08  1996-07-23   
7      10255      RICSU           9  1996-07-12   1996-08-09  1996-07-15   
8      10256      WELLI           3  1996-07-15   1996-08-12  1996-07-17   
9      10257      HILAA           4  1996-07-16   1996-08-13  1996-07-22   
10     10258      ERNSH           1  1996-07-17   1996-08-14  1996-07-23   
11     10259      CENTC           4  1996-07-18   1996-08-15  1996-07-25   
12     10260      OTTIK           4  1996-07-19   1996-08-16  1996-07-29   
13     10261      QUEDE           4  1996-07-19   1996-08-16  1996-07-30   
14     10262      RATTC           8  1996-07-22   1996-08-19  1996-07-25   
15     10263      ERNSH           9  1996-07-23   1996-08-20  1996-07-31   
16     10264      FOLKO           6  1996-07-24   1996-08-21  1996-08-23   
17     10265      BLONP           2  1996-07-25   1996-08-22  1996-08-12   
18     10266      WARTH           3  1996-07-26   1996-09-06  1996-07-31   
19     10267      FRANK           4  1996-07-29   1996-08-26  1996-08-06   
20     10268      GROSR           8  1996-07-30   1996-08-27  1996-08-02   
21     10269      WHITC           5  1996-07-31   1996-08-14  1996-08-09   
22     10270      WARTH           1  1996-08-01   1996-08-29  1996-08-02   
23     10271      SPLIR           6  1996-08-01   1996-08-29  1996-08-30   
24     10272      RATTC           6  1996-08-02   1996-08-30  1996-08-06   
25     10273      QUICK           3  1996-08-05   1996-09-02  1996-08-12   
26     10274      VINET           6  1996-08-06   1996-09-03  1996-08-16   
27     10275      MAGAA           1  1996-08-07   1996-09-04  1996-08-09   
28     10276      TORTU           8  1996-08-08   1996-08-22  1996-08-14   
29     10277      MORGK           2  1996-08-09   1996-09-06  1996-08-13   
..       ...        ...         ...         ...          ...         ...   
800    11048      BOTTM           7  1998-04-24   1998-05-22  1998-04-30   
801    11049      GOURL           3  1998-04-24   1998-05-22  1998-05-04   
802    11050      FOLKO           8  1998-04-27   1998-05-25  1998-05-05   
803    11051      LAMAI           7  1998-04-27   1998-05-25        None   
804    11052      HANAR           3  1998-04-27   1998-05-25  1998-05-01   
805    11053      PICCO           2  1998-04-27   1998-05-25  1998-04-29   
806    11054      CACTU           8  1998-04-28   1998-05-26        None   
807    11055      HILAA           7  1998-04-28   1998-05-26  1998-05-05   
808    11056      EASTC           8  1998-04-28   1998-05-12  1998-05-01   
809    11057      NORTS           3  1998-04-29   1998-05-27  1998-05-01   
810    11058      BLAUS           9  1998-04-29   1998-05-27        None   
811    11059      RICAR           2  1998-04-29   1998-06-10        None   
812    11060      FRANS           2  1998-04-30   1998-05-28  1998-05-04   
813    11061      GREAL           4  1998-04-30   1998-06-11        None   
814    11062      REGGC           4  1998-04-30   1998-05-28        None   
815    11063      HUNGO           3  1998-04-30   1998-05-28  1998-05-06   
816    11064      SAVEA           1  1998-05-01   1998-05-29  1998-05-04   
817    11065      LILAS           8  1998-05-01   1998-05-29        None   
818    11066      WHITC           7  1998-05-01   1998-05-29  1998-05-04   
819    11067      DRACD           1  1998-05-04   1998-05-18  1998-05-06   
820    11068      QUEEN           8  1998-05-04   199

### 5.a: How many orders in total?

In [39]:
orders.shape[0]

830

### 5.b: Which country is receiving the most orders?

In [40]:
orders[['OrderID','ShipCountry']].groupby(by = 'ShipCountry').count().sort_values(by = 'OrderID', ascending = False).head()

OrderID
ShipCountry         
USA              122
Germany          122
Brazil            83
France            77
UK                56

Answer: The USA and Germany each receive 122 orders.

### 5.c: Which country is receiving the least?

In [41]:
orders[['OrderID','ShipCountry']].groupby(by = 'ShipCountry').count().sort_values(by = 'OrderID', ascending = True).head()

OrderID
ShipCountry         
Norway             6
Poland             7
Portugal          13
Argentina         16
Switzerland       18

Answer: Norway receives the fewest orders with 6.

### 5.d: What's the average shipping time (ShippedDate - OrderDate)?

In [42]:
(orders['ShippedDate'] - orders['OrderDate']).mean()

Timedelta('8 days 11:48:25.809641')

Answer: The average shipping time looks to be about eight days, eleven hours, and 48 minutes.

### 5.e: What customer is submitting the highest number of orders?

In [43]:
orders[['CustomerID','OrderID']].groupby(by = 'CustomerID').count().sort_values(by = 'OrderID', ascending = False).head()

OrderID
CustomerID         
SAVEA            31
ERNSH            30
QUICK            28
HUNGO            19
FOLKO            19

Answer: CustomerID SAVEA is submitting the highest number of orders.